engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/formatter/ConfluentAvroToJsonFormatter.scala

```diff
@@ -24,64 +29,62 @@ private[confluent] class ConfluentAvroToJsonFormatter(schemaRegistryClientFactor
-    if (formatKey) {
-      printId(record.key(), printStream)
-    }
-    printId(record.value(), printStream)
-    if (formatKey) {
-      formatter.writeTo(record.key(), printStream)
-      printStream.print(Separator)
-    }
-    formatter.writeTo(record.value(), printStream)
-    bos.toByteArray
-  }
+    val deserializedRecord: ConsumerRecord[K, V] = deserializationSchema.deserialize(record)
+
+    val (keySchemaIdOpt, keyJsonOpt) = determineSchemaIdAndContent[K](deserializedRecord.key(), record.key(), keySchemaDataOpt)
+    val (valueSchemaIdOpt, valueJsonOpt) = determineSchemaIdAndContent[V](deserializedRecord.value(), record.value(), valueSchemaDataOpt)
 
-  private def printId(bytes: Array[Byte], printStream: PrintStream): Unit = {
-    val id = ConfluentUtils.readId(bytes)
-    printStream.print(id)
-    printStream.print(Separator)
+    formatter.serializeRecord(deserializedRecord, keySchemaIdOpt, keyJsonOpt, valueSchemaIdOpt, valueJsonOpt)
   }
 
   override def parseRecord(topic: String, formatted: Array[Byte]): ConsumerRecord[Array[Byte], Array[Byte]] = {
-    val str = new String(formatted, StandardCharsets.UTF_8)
-    val (keySchema, valueSchema, remainingString) = if (formatKey) {
-      val (ks, valueSchemaIdAndRest) = readSchemaId(str)
-      val (vs, rs) = readSchemaId(valueSchemaIdAndRest)
-      (ks, vs, rs)
-    } else {
-      val (vs, rs) = readSchemaId(str)
-      (null, vs, rs)
-    }
-    reader(topic).readMessage(remainingString, keySchema, valueSchema)
+    val serializableConsumerRecord: SerializableConsumerRecord[Json, Json] = reader.toSerializable(formatted)
+
+    serializableConsumerRecord.toKafkaConsumerRecord(topic, serializeKeyValue)
   }
 
-  private def readSchemaId(str: String): (Schema, String) = {
-    val separatorIndx = str.indexOf(Separator)
-    if (separatorIndx < 1)
-      throw new IllegalStateException(s"Cannot find schema id separtor: $Separator in text: $str")
-    val id = Integer.parseInt(str.substring(0, separatorIndx))
-    val remaining = if (separatorIndx + 1 > str.length) "" else str.substring(separatorIndx + 1)
-    val parsedSchema = schemaRegistryClient.client.getSchemaById(id)
-    val schema = ConfluentUtils.extractSchema(parsedSchema)
-    (schema, remaining)
+  private def determineSchemaIdAndContent[T: ClassTag](obj: T, rawData: Array[Byte], schemaDataOpt: Option[RuntimeSchemaData]): (Option[Int], Option[Json]) = {
+    schemaDataOpt match {
+      case Some(RuntimeSchemaData(_, _, Some(runtimeFallbackStrategy))) =>
+        // fallback
+        val jsonOpt = Some(runtimeFallbackStrategy.toJson(obj.asInstanceOf[runtimeFallbackStrategy.T]))
+        (None, jsonOpt)
+      case _ if AvroUtils.isSpecificRecord[T] =>
+        // SpecificRecord
+        val schemaIdOpt = schemaDataOpt.flatMap(_.schemaIdOpt)
+        val jsonOpt = schemaDataOpt.map(runtimeSchema => formatter.specificToJson(obj.asInstanceOf[AnyRef], runtimeSchema.schema))
```

Comment:
instead of dispatching between specific record/generic record case here, it would be nice to have `AvroUtils.createDatumWriter` that will do this dispatching on their own.

Reply:
Good point, it is moved to separate PR.